pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -1447,6 +1447,243 @@ func TestListUsersEdgePruning(t *testing.T) {
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: `model
+            schema 1.1
+          type user
+          type document
+            relations
+              define public_1: [user:*,user]
+			  define public_2: [user:*,user]
+              define is_public: public_1 and public_2`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "public_1", "user:maria"),
+				tuple.NewTupleKey("document:1", "public_2", "user:maria"),
+
+				tuple.NewTupleKey("document:1", "public_1", "user:*"),
+				tuple.NewTupleKey("document:1", "public_2", "user:*"),
+
+				tuple.NewTupleKey("document:1", "public_1", "user:jon"),
+			},
+			expectedUsers: []string{"user:maria", "user:*", "user:jon"},
+		},
+		{
+			name: "with_multiple_wildcards_2",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "can_view",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: `model
+		    schema 1.1
+		  type user
+		  type document
+		    relations
+		      define viewer: [user:*,user]
+		      define this_is_not_assigned_to_any_user: [user]
+		      define can_view: viewer and this_is_not_assigned_to_any_user`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "viewer", "user:*"),
+				tuple.NewTupleKey("document:1", "viewer", "user:will"),
+			},
+			expectedUsers: []string{},
+		},
+		{
+			name: "with_multiple_wildcards_3",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "can_view",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: `model
+		    schema 1.1
+		  type user
+		  type document
+		    relations
+		      define viewer: [user:*,user]
+		      define required: [user:*,user]
+		      define can_view: viewer and required`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "viewer", "user:*"),
+				tuple.NewTupleKey("document:1", "required", "user:*"),
+				tuple.NewTupleKey("document:1", "viewer", "user:will"),
+				tuple.NewTupleKey("document:1", "required", "user:will"),
+			},
+			expectedUsers: []string{"user:*", "user:will"},
+		},
+		{
+			name: "with_multiple_wildcards_4",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "can_view",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: `model
+		    schema 1.1
+		  type user
+		  type document
+		    relations
+		      define required_1: [user]
+		      define required_2: [user]
+		      define can_view: required_1 and required_2`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "required_1", "user:*"),
+				tuple.NewTupleKey("document:1", "required_2", "user:*"),
```

Comment:
these two tuples aren't allowed by the model you defined

Reply:
That's exactly what I'm trying to test. Or is it safe to assume that this is disallowed by the datastore?

Reply:
The datastore doesn't check for correctness, the API has to do that. So you have make sure that the API filters out invalid tuples with respect to the model being used. There are two approaches for this:

1. Use a filtered tuple iterator `filteredIter := storage.NewFilteredTupleKeyIterator(storage.NewTupleKeyIteratorFromTupleIterator(tupleIter), validation.FilterInvalidTuples(typesys))`
2. Use the graph of the model to pre-filter what tuples to fetch (like we do in ReverseExpand)

Reply:
We currently employ strategy 1. This test is ensuring that it works correctly. I have a strong preference to maintain this test case because it is easily forgotten and can wreak havoc on our system.

Reply:
Ok, then can you please leave a comment next to each "invalid" tuple ?